File: shlomif-c-snippets/calc-pi/multi-digits.cpp
#include <malloc.h>
#include <string.h>
#include <stdio.h>

const int base = 10000;
typedef signed short multi_digit_t;

class MultiDigit
{
public:
    multi_digit_t * digits;
    int length;
public:
    MultiDigit();
    MultiDigit(MultiDigit &);
    MultiDigit(unsigned long long);
    ~MultiDigit()  { free(digits); }
    int from_number(unsigned long long number);
    void print(FILE *);
    void copy(MultiDigit & a);
    int from_string(char * s);
    int from_hexstring(char * s);
};

MultiDigit::MultiDigit()
{
    digits = (multi_digit_t*)malloc(sizeof(multi_digit_t)*1);
    digits[0] = 0;
    length = 1;
}

MultiDigit::MultiDigit(unsigned long long value)
{
    digits  = NULL;
    from_number(value);
}

MultiDigit::MultiDigit(MultiDigit & a)
{
    digits = NULL;
    copy(a);
}

void MultiDigit::copy(MultiDigit & a)
{
    length = a.length;
    digits = (multi_digit_t*)realloc(digits, a.length*sizeof(multi_digit_t));
    memcpy(digits, a.digits, a.length*sizeof(multi_digit_t));
}

int MultiDigit::from_number(unsigned long long number)
{
    int a;
    unsigned long long n = number;
    a=0;

    while (n > 0)
    {
        a++;
        n = n / base;
    }
    digits = (multi_digit_t*)realloc(digits, sizeof(multi_digit_t)*a);

    a=0;
    while (number > 0)
    {
        digits[a] = (multi_digit_t)(number % base);
        a++;
        number = number / base;
    }
    length = a;
    return 1;
}

void MultiDigit::print(FILE * f = stdout)
{
    int a;
    for(a=0;a<length;a++)
    {
        fprintf(f, (a==0)?"%i,":"%04i,", digits[length-a-1]);
    }
    printf("%s", "\n");
}

int multiply_multi_digit(MultiDigit & one, MultiDigit & two, MultiDigit * result);
int add_multi_digit(MultiDigit & one, MultiDigit & two, MultiDigit * result);

int MultiDigit::from_string(char * s)
{
    if (!strncmp(s,"0x",2))
    {
        // return from_hexstring(s+2);
    }
    else
    {
        MultiDigit d, accum2(0), ten(10);

        from_number(0);

        while (((*s) >= '0') && ((*s) <= '9'))
        {
            d.from_number(*s-'0');
            multiply_multi_digit(*(this), ten, &accum2);
            add_multi_digit(accum2, d, this);
        }
    }
}

int add_multi_digit(MultiDigit & one, MultiDigit & two, MultiDigit * result)
{
    int min_len, max_len, a, sum, which;
    if (one.length > two.length)
    {
        max_len = one.length;
        min_len = two.length;
        which = 0;
    }
    else
    {
        min_len = one.length;
        max_len = two.length;
        which = 1;
    }

    result->length = max_len+1;
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));
    for(a=0;a<max_len+1;a++)
    {
        result->digits[a] = 0;
    }
    for(a=0;a<min_len;a++)
    {
        sum = one.digits[a] + two.digits[a] + result->digits[a];
        result->digits[a] = sum % base;
        result->digits[a+1] = sum / base;
    }
    for(;a<max_len;a++)
    {
        if (which == 0)
        {
            sum = one.digits[a] + result->digits[a];
        }
        else
        {
            sum = two.digits[a] + result->digits[a];
        }
        result->digits[a] = sum % base;
        result->digits[a+1] = sum / base;
    }

    while (result->digits[result->length-1] == 0)
    {
        result->length--;
        if (result->length == 1)
            break;
    }
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));


    return 1;
}

int subtract_multi_digit(MultiDigit & one, MultiDigit & two, MultiDigit * result)
{
    int min_len, max_len, a, diff;
    if (one.length > two.length)
    {
        max_len = one.length;
        min_len = two.length;
    }
    else
    {
        min_len = one.length;
        max_len = two.length;
    }

    result->length = max_len;
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));
    for(a=0;a<max_len;a++)
    {
        result->digits[a] = 0;
    }
    for(a=0;a<max_len;a++)
    {
        diff = result->digits[a];
        if (one.length > a)
        {
            diff += one.digits[a];
        }
        if (two.length > a)
        {
            diff -= two.digits[a];
        }
        if (diff >= 0)
        {
            result->digits[a] = diff;
        }
        else
        {
            result->digits[a] = diff+base;
            result->digits[a+1] = -1;
        }
    }

    while (result->digits[result->length-1] == 0)
    {
        result->length--;
        if (result->length == 1)
            break;
    }
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));


    return 1;
}


int multiply_multi_digit(MultiDigit & one, MultiDigit & two, MultiDigit * result)
{
    int a,b,c, product;

    result->length = one.length+two.length;
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));
    for(a=0;a<result->length;a++)
    {
        result->digits[a]=0;
    }
    for(a=0;a<one.length;a++)
    {
        for(b=0;b<two.length;b++)
        {
            product = one.digits[a] * two.digits[b];
            result->digits[a+b] += product % base;
            for(c=a+b;result->digits[c]>=base;c++)
            {
                result->digits[c+1] += result->digits[c] / base;
                result->digits[c] %= base;
            }
            if ((product / base) > 0)
            {
                result->digits[a+b+1] += (product / base);
                for(c=a+b+1;result->digits[c]>=base;c++)
                {
                    result->digits[c+1] += (result->digits[c] / base);
                    result->digits[c] %= base;
                }
            }
        }
    }

    while (result->digits[result->length-1] == 0)
    {
        result->length--;
        if (result->length == 1)
            break;
    }
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));

    return 1;
}

#define max(a,b) ((a) > (b) ? (a) : (b))

int divide_multi_digit(MultiDigit & divide_, MultiDigit & div_by, MultiDigit * result)
{
    MultiDigit divide[2];
    int a, d_digit, db_digit, offset, multiply_by;
    int which=0;
    MultiDigit product;

    divide[which].copy(divide_);      // clone it because we are going to change it

    result->length = max(divide[which].length-div_by.length+1, 1);
    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));
    for(a=0;a < result->length; a++)
    {
        result->digits[a] = 0;
    }

    d_digit = divide[which].length-1;
    db_digit = -1;
    for(a=div_by.length-1;a>=0;a--)
    {
        if (div_by.digits[a] > 0)
        {
            db_digit = a;
            break;
        }
    }
    if (db_digit == -1) // div_by equals to 0
    {
        return 0;
    }

    while (1)
    {
        // Check if the division can be ended
        for(a=d_digit, d_digit = -1;a>=0;a--)
        {
            if (divide[which].digits[a] > 0)
            {
                d_digit = a;
                break;
            }
        }

        if (db_digit > d_digit)
        {
            while (result->digits[result->length-1] == 0)
            {
                result->length--;
                if (result->length == 1)
                    break;
            }
            result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));


            return 1;
        }


        if (db_digit == d_digit)
        {
            for(a=d_digit;a>=0;a--)
            {
                if (div_by.digits[a] > divide[which].digits[a])
                {
                    while (result->digits[result->length-1] == 0)
                    {
                        result->length--;
                        if (result->length == 1)
                            break;
                    }
                    result->digits = (multi_digit_t*)realloc(result->digits, result->length*sizeof(multi_digit_t));

                    return 1;
                }
                else if (div_by.digits[a] < divide[which].digits[a])
                {
                    break;
                }
            }
        }


        for (a=0; a<=db_digit; a++)
        {
            if (divide[which].digits[d_digit-a] != div_by.digits[db_digit-a])
            {
                break;
            }
        }

        if (a == db_digit+1)
        {
            multiply_by = 1;
            offset = d_digit - db_digit;
        }
        else if (divide[which].digits[d_digit-a] > div_by.digits[db_digit-a])
        {
            offset = d_digit - db_digit;
            if (a>0)
            {
                multiply_by = 1;
            }
            else
            {
                int b = ((divide[which].digits[d_digit]*base + ((d_digit>0)?divide[which].digits[d_digit-1]:0)) /
                             (div_by.digits[db_digit]*base + ((db_digit>0)?div_by.digits[db_digit-1]:0)));
                multiply_by = (b>1)?(b-1):b;
            }
        }
        else
        {
            offset = d_digit - db_digit - 1;
            if (a>1)
            {
                multiply_by = base-1;
            }
            else
            {
                int b = (multi_digit_t) ((((long long)divide[which].digits[d_digit])*base*base + ((d_digit>0)?divide[which].digits[d_digit-1]:0)*base +  ((d_digit>1)?divide[which].digits[d_digit-1]:0)) /
                             (div_by.digits[db_digit]*base + ((db_digit>0)?div_by.digits[db_digit-1]:0)));
                multiply_by = (b>1)?(b-1):b;
            }
        }
        {
            MultiDigit m(multiply_by);
            multiply_multi_digit(div_by, m, &product);

            product.length += offset;
            product.digits = (multi_digit_t*)realloc(product.digits, sizeof(multi_digit_t)*product.length);
            memmove(product.digits+offset,product.digits,sizeof(multi_digit_t)*(product.length-offset));
            for(a=0;a<offset;a++)
            {
                product.digits[a] = 0;
            }
        }
        subtract_multi_digit(divide[which], product, &divide[1-which]);
        which = 1-which;
        result->digits[offset] += multiply_by;
        for(a=offset;result->digits[a]>=base;a++)
        {
            result->digits[a+1] += result->digits[a] / base;
            result->digits[a] %= base;
        }
    }
}

#if 0

int main()
{
    MultiDigit a(1390), b(49);
    MultiDigit c;

    divide_multi_digit(a,b,&c);
    c.print();

    return 0;
}

#endif

int main()
{
    MultiDigit numer(2), numer2, denom[2], item[2], pi, m, m2;
    item[0].from_number(2);
    denom[0].from_number(1);
    int which=0;
    int a;

    //for(a=1;a<=33250;a++)
    for(a=1;a<=3350;a++)
    {
        if (a % 100 == 0)
        {
            printf("%i\n", a);
        }
        m.from_number(2*a+1);
        multiply_multi_digit(denom[which], m, &denom[1-which]);
        m2.from_number(a);
        multiply_multi_digit(item[which], m2, &item[1-which]);
        multiply_multi_digit(numer, m, &numer2);
        add_multi_digit(item[1-which], numer2, &numer);
        which = 1-which;
    }

    int offset = 250;

    numer.length += offset;
    numer.digits = (multi_digit_t*)realloc(numer.digits, sizeof(multi_digit_t)*numer.length);
    memmove(numer.digits+offset,numer.digits,sizeof(multi_digit_t)*(numer.length-offset));
    for(a=0;a<offset;a++)
    {
        numer.digits[a] = 0;
    }

    divide_multi_digit(numer, denom[which], &pi);

    FILE * f = fopen("dump.txt", "w");

    printf("pi is ");
    pi.print();
    fprintf(f,"pi is ");
    pi.print(f);
    fclose(f);

    return 0;
}

File: shlomif-c-snippets/calc-pi/multi-digits.pl

use strict;
use warnings;

my $base = 10000;

#my $base = 10;

sub _base_moddiv
{
    my ($n) = @_;
    return ( $n % $base, int( $n / $base ) );
}

{
    my @r = _base_moddiv( 9 * $base + 5 );
    if ( $r[0] != 5 )
    {
        die;
    }
    if ( $r[1] != 9 )
    {
        die;
    }
}

sub _trim_zeroes
{
    my ($result) = @_;
    while ( @$result > 1 and ( $result->[-1] == 0 ) )
    {
        pop @$result;
    }

    return;
}

sub add_multi_digit
{
    my ( $one, $two ) = @_;
    my ( $ii, );
    if ( scalar( @{$one} ) < scalar( @{$two} ) )
    {
        ( $one, $two ) = ( $two, $one );
    }
    my $max_len = scalar( @{$one} );
    my $min_len = scalar( @{$two} );

    my @result = ( (0) x ( $max_len + 1 ) );
    for ( $ii = 0 ; $ii < $min_len ; ++$ii )
    {
        my $sum = $one->[$ii] + $two->[$ii] + $result[$ii];
        @result[ $ii, $ii + 1 ] = _base_moddiv($sum);
    }
    for ( ; $ii < $max_len ; ++$ii )
    {
        my $sum = $one->[$ii] + $result[$ii];
        @result[ $ii, $ii + 1 ] = _base_moddiv($sum);
    }
    _trim_zeroes( \@result );
    return \@result;
}

sub subtract_multi_digit
{
    my ( $one, $two ) = @_;
    my ( $min_len, $max_len, );
    if ( scalar( @{$one} ) > scalar( @{$two} ) )
    {
        $max_len = scalar( @{$one} );
        $min_len = scalar( @{$two} );
    }
    else
    {
        $min_len = scalar( @{$one} );
        $max_len = scalar( @{$two} );
    }

    my @result = ( (0) x ($max_len) );
    for my $ii ( 0 .. $max_len - 1 )
    {
        my $diff = $result[$ii];
        if ( scalar( @{$one} ) > $ii )
        {
            $diff += $one->[$ii];
        }
        if ( scalar( @{$two} ) > $ii )
        {
            $diff -= $two->[$ii];
        }
        if ( $diff >= 0 )
        {
            $result[$ii] = $diff;
        }
        else
        {
            $result[$ii] = $diff + $base;
            $result[ $ii + 1 ] = -1;
        }
    }

    _trim_zeroes( \@result );
    return \@result;
}

sub multiply_multi_digit
{
    my ( $one, $two ) = @_;
    my ( $ii, $jj, $product, $c );
    my @result = ( (0) x ( scalar( @{$one} ) + scalar( @{$two} ) ) );
    for ( $ii = 0 ; $ii < scalar( @{$one} ) ; ++$ii )
    {
        for ( $jj = 0 ; $jj < scalar( @{$two} ) ; $jj++ )
        {
            $product = $one->[$ii] * $two->[$jj];
            $result[ $ii + $jj ] += $product % $base;
            for ( $c = $ii + $jj ; $result[$c] >= $base ; $c++ )
            {
                $result[ $c + 1 ] += int( $result[$c] / $base );
                $result[$c] %= $base;
            }
            if ( int( $product / $base ) > 0 )
            {
                $result[ $ii + $jj + 1 ] += int( $product / $base );
                for ( $c = $ii + $jj + 1 ; $result[$c] >= $base ; $c++ )
                {
                    $result[ $c + 1 ] += int( $result[$c] / $base );
                    $result[$c] %= $base;
                }
            }
        }
    }

    _trim_zeroes( \@result );

    return \@result;
}

sub divide_multi_digit
{
    my ( $divide, $div_by ) = @_;
    $divide = [ @{$divide} ];    # clone it because we are going to change it
    my ( @result, $ii, $d_digit, $db_digit, $offset, $multiply_by, @product );

    $result[0] = 0;
    for (
        $ii = 0 ;
        $ii < scalar( @{$divide} ) - scalar( @{$div_by} ) + 1 ;
        ++$ii
        )
    {
        $result[$ii] = 0;
    }

    $d_digit  = scalar( @{$divide} ) - 1;
    $db_digit = -1;
    for ( $ii = scalar( @{$div_by} ) - 1 ; $ii >= 0 ; --$ii )
    {
        if ( $div_by->[$ii] > 0 )
        {
            $db_digit = $ii;
            last;
        }
    }
    if ( $db_digit == -1 )    # div_by equals to 0
    {
        return undef;
    }

    while (1)
    {
        if ( $d_digit > $#$divide )
        {
            $d_digit = $#$divide;
        }

        # Check if the division can be ended
        for ( $ii = $d_digit, $d_digit = -1 ; $ii >= 0 ; --$ii )
        {
            if ( $divide->[$ii] > 0 )
            {
                $d_digit = $ii;
                last;
            }
        }

        if ( $db_digit > $d_digit )
        {
            _trim_zeroes( \@result );
            return \@result;
        }

        if ( $db_digit == $d_digit )
        {
            for ( $ii = $d_digit ; $ii >= 0 ; --$ii )
            {
                if ( $div_by->[$ii] > $divide->[$ii] )
                {
                    _trim_zeroes( \@result );

                    return \@result;
                }
                elsif ( $div_by->[$ii] < $divide->[$ii] )
                {
                    last;
                }
            }
        }

        for ( $ii = 0 ; $ii <= $db_digit ; ++$ii )
        {
            if ( $divide->[ $d_digit - $ii ] != $div_by->[ $db_digit - $ii ] )
            {
                last;
            }
        }

        if ( $ii == $db_digit + 1 )
        {
            $multiply_by = 1;
            $offset      = $d_digit - $db_digit;
        }
        elsif ( $divide->[ $d_digit - $ii ] > $div_by->[ $db_digit - $ii ] )
        {
            $offset = $d_digit - $db_digit;
            if ( $ii > 0 )
            {
                $multiply_by = 1;
            }
            else
            {
                my $jj = int(
                    (
                        $divide->[$d_digit] * $base + (
                            ( $d_digit > 0 ) ? $divide->[ $d_digit - 1 ] : 0
                        )
                    ) / (
                        $div_by->[$db_digit] * $base + (
                            ( $db_digit > 0 ) ? $div_by->[ $db_digit - 1 ] : 0
                        )
                    )
                );
                $multiply_by = ( $jj > 1 ) ? ( $jj - 1 ) : $jj;
            }
        }
        else
        {
            $offset = $d_digit - $db_digit - 1;
            if ( $ii > 1 )
            {
                $multiply_by = $base - 1;
            }
            else
            {
                my $jj = int(
                    (
                        $divide->[$d_digit] * $base * $base + (
                            ( $d_digit > 0 ) ? $divide->[ $d_digit - 1 ] : 0
                        ) * $base + (
                            ( $d_digit > 1 ) ? $divide->[ $d_digit - 1 ] : 0
                        )
                    ) / (
                        $div_by->[$db_digit] * $base + (
                            ( $db_digit > 0 ) ? $div_by->[ $db_digit - 1 ] : 0
                        )
                    )
                );
                $multiply_by = ( $jj > 1 ) ? ( $jj - 1 ) : $jj;
            }
        }
        @product = @{ multiply_multi_digit( $div_by, [$multiply_by] ) };
        @product = ( ( ( 0, ) x $offset ), @product );
        $divide  = subtract_multi_digit( $divide, \@product );
        $result[$offset] += $multiply_by;
        for ( $ii = $offset ; $result[$ii] >= $base ; ++$ii )
        {
            $result[ $ii + 1 ] += int( $result[$ii] / $base );
            $result[$ii] %= $base;
        }
    }
}

sub print_multi_digit
{
    my $number = shift(@_);
    my $ii;
    for ( $ii = 0 ; $ii < scalar( @{$number} ) ; ++$ii )
    {
        printf(
            ( $ii == 0 ) ? "%i," : "%04i,",
            $number->[ scalar( @{$number} ) - $ii - 1 ]
        );

        #print $number->[scalar(@{$number})-$ii-1];
    }
    print "\n";
}

sub to_multi_digit
{
    my $number = shift(@_);
    my @result;
    while ( $number > 0 )
    {
        push @result, $number % $base;
        $number = int( $number / $base );
    }
    return \@result;
}

my $numer = [2];
my $item  = [2];
my $denom = [1];

for my $i ( 1 .. 130 )
{
    if ( $i % 30 == 0 )
    {
        print "$i\n";
    }
    my $factor = to_multi_digit( 2 * $i + 1 );
    $denom = multiply_multi_digit( $denom, $factor, );
    $item  = multiply_multi_digit( $item,  to_multi_digit($i) );
    $numer = add_multi_digit( $item, multiply_multi_digit( $numer, $factor, ) );

    #print "denom is ", join('', reverse(@{$denom})), "\n";
    #print "item is ", join('', reverse(@{$item})), "\n";
    #print "numer is ", join('', reverse(@{$numer})), "\n";
}

my $pi     = divide_multi_digit( [ ( ( 0, ) x 40 ), @{$numer} ], $denom );
my $pi_str = join( '', reverse( @{$pi} ) );
if ( $pi_str !~ m#\A3(?:\.)?1415926535#ms )
{
    die qq#"$pi_str" has wrong digits!#;
}
my $msg = "pi is " . $pi_str . "\n";
open my $out_fh, '>', 'dump.txt';
print $msg;
print {$out_fh} $msg;
close($out_fh);

__END__

=head1 COPYRIGHT & LICENSE

Copyright 2023 by Shlomi Fish

This program is distributed under the MIT / Expat License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut
